### PR TITLE
Implement "Users" bounded context

### PR DESCRIPTION
**Pull request by vlad-lubenskyi**
users 2

This PR:
 - upgrades `gradlew` and and Gradle configuration;
 - renames module `core` to `users`;
 - introduces `Users` bounded context and the process of sign-in.

## `User`

A user is a member of 0+ `Groups`. A user has explicit roles assigned directly, and implicit roles propagated from groups.

A user can be a part of an organizational structure and may, therefore, belong to either `OrgUnit` or `Organization`.

The aggregate can be extended with additional information using `UserAttribute`s.

## `Group`

A group of users and groups. A group has explicit roles assigned directly (they're propagated further), and implicit roles propagated from parent groups. 

A group can be a part of an organizational structure and may, therefore, belong to either `OrgUnit` or `Organization`.

The aggregate can be extended with additional information using `GroupAttribute`s.

## `OrgUnit`

An organizational unit. An organizational unit must be a part of `Organization` or a larger `OrgUnit`.

The aggregate can be extended with additional information using `OrgUnitAttribute`s.

## `Organization`

An organization is a root of the organizational structure of the customer. An organization is associated with the tenant.

The aggregate can be extended with additional information using `OrganizationAttribute`s.

## `Role`

A role belongs either to `Organization` or `OrgUnit` and can be assigned to users and groups from that unit.

## `RemoteIdentitySignIn` process manager

The process manager that perform sign in by querying remote identity providers. Communication with identity providers is delegated to `RemoteIdentityProvider` process manager.

## `RemoteIdentityProvider` process manager

The abstract process manager whose sole responsibility is to fetch information from the identity provider.

### CODE REVIEW

**Comment by vlad-lubenskyi** [comment]
@armiol PTAL

**Comment by vlad-lubenskyi** [comment]
@armiol PTAL

File: users/src/main/proto/spine/users/c/group/group.proto [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+message Group {
+    option (entity).kind = AGGREGATE;
+    option (required_field) = "unit|organization";
+
+    GroupId id = 1;
+
+    // The name of a group to display on UI.
+    string display_name = 2 [(required) = true];
+
+    // An interent domain of the group.
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/group/group.proto]: Please have a comment on why this field is optional. Most likely, you'll also need to add an example to the message-level documentation, explaining what kind of groups there may be. And include a sample with/without an internet domain.

File: users/src/main/proto/spine/users/c/group/group_attribute.proto [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+// An intrinsic attribute of a group.
+//
+// Adding attributes is the way to extend `Group` to meet the specific needs of the application.
+//
+// For example, the following pieces of information make good attributes:
+//
+// * { name: "description", value: "People who can access printer on the 2nd floor" }
+//
+message GroupAttribute {
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/group/group_attribute.proto]: We need to have a comment on why `GroupAttribute` and `OrgAttribute` are different messages with the same content.

Or have just a single `Attribute` message.

File: users/src/main/proto/spine/users/c/group/commands.proto [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+// The command to update a group attribute.
+//
+message UpdateGroupAttribute {
+
+    option (events) = "GroupAttributeUpdated";
+
+    GroupId id = 1;
+
+    // The new value of an attribute.
+    GroupAttribute attribute = 2 [(required) = true];
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/group/commands.proto]: So basically we need to explain (in `Group` and here?) that group attributes should have unique names per `Group` instance.

Same for `Organization`.

I think we may use `map<string, ... Any>` in data structures and just have 
```
// in commands

string name = 2 [(required) = true];
Any new_value = 3;

// and in events:

string name = 2 [(required) = true];
Any new_value = 3;
Any old_value = 4;
```
thus getting rid of several redundant messages to maintain.

File: users/src/main/proto/spine/users/c/group/events.proto [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+//
+message GroupCreated {
+    option (required_field) = "unit|organization";
+
+    GroupId id = 1;
+
+    // The name of a group to display on UI.
+    string display_name = 2 [(required) = true];
+
+    // The roles assigned to the user directly (aka not propagated by a group membership).
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/group/events.proto]: Who is `user`?

File: users/src/main/proto/spine/users/c/role/events.proto [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+        OrgUnitId org_unit = 4;
+    }
+}
+
+// The event is emitted when a role is deleted.
+//
+message RoleDeleted {
+
+    OrgUnitId id = 1;
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/role/events.proto]: I don't think that's true.

File: users/src/main/proto/spine/users/c/role/role.proto [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+option java_package = "io.spine.users.role";
+option java_outer_classname = "RoleProto";
+option java_multiple_files = true;
+option java_generate_equals_and_hash = true;
+
+import "spine/users/identifiers.proto";
+
+// A role in the system.
+//
+// Roles can be assigned to users and groups.
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/role/role.proto]: There is so much more to tell — for instance how roles that are organization-wide talk to roles in the organizational units. 

Please describe the use cases and propagation rules.

File: users/src/main/proto/spine/users/c/signin/events.proto [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import "spine/users/user_auth_identity.proto";
+
+// The event is emitted when the sign-in process is successfully finished.
+//
+message RemoteIdentitySignInFinished {
+
+    spine.core.UserId id = 1;
+
+    // The authentication identity used to sign-in.
+    UserAuthIdentity identity = 2;
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/events.proto]: I think this field is required. Here and below.

File: users/src/main/proto/spine/users/c/signin/identity/commands.proto [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+// The command to check if user account is active and thus can be signed in.
+//
+message CheckUserStatus {
+    option (events) = "UserStatusChecked";
+    option (rejections) = "RemoteIdentityNotFound";
+
+    RemoteIdentityProviderId id = 1;
+
+    // The ID of a user to check status.
+    spine.core.UserId user_id = 2;
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/identity/commands.proto]: It's also `required`.  Also here and in the message below.

File: users/src/main/proto/spine/users/c/signin/identity/commands.proto [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+option java_multiple_files = true;
+option java_generate_equals_and_hash = true;
+
+import "spine/core/user_id.proto";
+import "spine/users/identifiers.proto";
+import "spine/users/user_auth_identity.proto";
+
+// The command to check if user account is active and thus can be signed in.
+//
+message CheckUserStatus {
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/identity/commands.proto]: Maybe `CheckUserActive`?

File: users/src/main/proto/spine/users/c/signin/identity/commands.proto [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    option (events) = "UserStatusChecked";
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/identity/commands.proto]: and then `UserActiveStatusConfirmed`?

File: users/src/main/proto/spine/users/c/signin/identity/commands.proto [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    option (rejections) = "RemoteIdentityNotFound";
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/identity/commands.proto]: and probably `CannotSignInInactiveUser`?

File: users/src/main/proto/spine/users/c/signin/rejections.proto [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import "spine/core/user_id.proto";
+import "spine/users/user_auth_identity.proto";
+
+// This rejection is generated when a remote identity provider didn't find a requested identity.
+//
+message RemoteIdentityNotFound {
+
+    spine.core.UserId id = 1;
+
+    UserAuthIdentity identity = 2;
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/rejections.proto]: This is a required field.

File: users/src/main/proto/spine/users/c/signin/remote_identity_sign_in.proto [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+// On every sign-in, the process manager checks:
+//  * if the user exists in the system;
+//  * if the user is still active at the remote identity provider.
+//
+// If the user is not yet in the system, the process manager fetches the user profile from the
+// remote identity provider and creates the new `User` aggregate.
+//
+// If the user is active, the process ends with `UserSignInFinished` event. Otherwise, the process
+// ends with `UserSignInFailed` event. If necessary, the process manager synchronizes the user status
+// between TPIP and the `User` aggregate state.
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/remote_identity_sign_in.proto]: `TPIP` is something unfriendly to readers.

File: users/src/main/proto/spine/users/c/signin/remote_identity_sign_in.proto [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        // Indicate that the process manager is fetching the user profile.
+        USER_PROFILE_FETCHING = 2;
+
+        // Indicates that the user doesn't exist and the process manager is creating it.
+        USER_CREATING = 3;
+
+        // A user is signed-in.
+        SIGNED_IN = 4;
+
+        // The sign-in process failed and should be repeated.
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/signin/remote_identity_sign_in.proto]: Please explain in more details. There may be different types of failures. Some of them do not require repeating (unless an end-user wants to).

File: users/src/main/proto/spine/users/c/user/commands.proto [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+    spine.core.UserId id = 1;
+
+    // The new parent entity in the organizational structure.
+    ParentEntity new_parent_entity = 2 [(required) = true];
+}
+
+// The command to start a group membership.
+//
+message StartGroupMembership {
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/user/commands.proto]: `JoinGroup` and `UserJoinedGroup`

File: users/src/main/proto/spine/users/c/user/commands.proto [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+    spine.core.UserId id = 1;
+
+    // The ID of a group.
+    GroupId group_id = 2 [(required) = true];
+}
+
+// The command to stop a group membership.
+//
+message StopGroupMembership {
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/user/commands.proto]: `LeaveGroup` and `UserLeftGroup`

File: users/src/main/proto/spine/users/c/user/user.proto [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    // The attributes of the user.
+    repeated UserAttribute attribute = 6;
+
+    // General profile information for a user.
+    UserProfile profile = 7;
+
+    // The status of the user.
+    Status status = 8 [(required) = true];
+
+    // The primary authentication identity of the user.
```
Comment (armiol) [inline users/src/main/proto/spine/users/c/user/user.proto]: Please explain.

File: users/src/test/java/io/spine/users/signin/RemoteIdentitySignInPmEventTest.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+import io.spine.testing.server.procman.ProcessManagerEventReactionTest;
+
+import static io.spine.users.signin.given.SignInTestEnv.userId;
+
+/**
+ * An implementation base for the {@link RemoteIdentitySignInProcMan} event reactors tests.
+ *
+ * @author Vladyslav Lubenskyi
+ */
+abstract class RemoteIdentitySignInPmEventTest<E extends Message>
```
Comment (armiol) [inline users/src/test/java/io/spine/users/signin/RemoteIdentitySignInPmEventTest.java]: The generic parameter isn't documented.

**Pull request closed without merging**